### ISSUE TITLE
Build flow and compiling: Improving compilation by utilizing mingw/make – can't find SDL

### ISSUE THREAD

**Issue by RavenMacDaddy** [trzy/Supermodel#21]
@BlackPowerade and I are having a [discussion](https://github.com/Calinou/scoop-games/issues/681) over at the Scoop Games repository.

Scoop is a Command-Line Installer – more commonly known as a package manager.

We're trying to figure out how to improve the compiling of the emulator, particularly by using `mingw` or `make` rather than `python` – but so far we're unable to get those methods to find SDL in the process.

---

We're also trying to figure out how to make [regex](https://en.wikipedia.org/wiki/Regular_expression) recognize both the addition of a new compiled, and the hash of the corresponding build – but being able to do the compilation described above would supersede that.

**Comment by trzy** [comment]
The Python script is not intended for users to build the emulator, it’s just something I hacked together to run on a spare home PC as a build bot because SourceForge didn’t have infrastructure to do this and GitHub Actions’ documentation makes my eyes bleed.

The Python script invokes make via MSYS2’s bash. And this assumes that SDL2 is already installed by the user. The build system does not attempt to install any of the required dependencies. That has to be done manually before running make.

**Comment by RavenMacDaddy** [comment]
Good, that gives us a more informed starting point - cheers.

**Comment by BlackPowerade** [comment]
I already have the supporting libraries installed with mingw, including SDL. It even starts compiling.
The problem I am running into is getting GCC to include SDL. No matter what I set for 
`SDL_INCLUDE_DIR` and `SDL_LIB_DIR` in https://github.com/trzy/Supermodel/blob/master/Makefiles/Makefile.Win32#L52,L53   
It fails to find `SDL.h`.     
It throws this error everytime:
![image](https://user-images.githubusercontent.com/21957509/189768981-b419dd6d-26f2-4916-83ef-cb2f479e170c.png)
Yet I know gcc can find the includes, it looks in them earlier:     
![image](https://user-images.githubusercontent.com/21957509/189769191-838a75e0-affb-405a-b775-e773d5cc938c.png)
SDL's include folder is properly populated:     
![image](https://user-images.githubusercontent.com/21957509/189770034-d33e5995-a14b-4a7f-afe5-63c0861f003d.png)
And SDL files are present under libs:    
![image](https://user-images.githubusercontent.com/21957509/189770110-996a0285-eaa6-450d-bfd6-3eb466010fea.png)

It's possible to have supermodel be automatically built, I just need to figure out what the makefile wants.

**Comment by trzy** [comment]
Where exactly on your hard drive is SDL and what are you setting SDL_INCLUDE_DIR and SDL_LIB_DIR to?

**Comment by BlackPowerade** [comment]
D:\msys64\mingw64\include\SDL2     
D:\msys64\mingw64\lib

**Comment by trzy** [comment]
That looks correct. What if you try stripping d: out and make sure you build somewhere on your d: drive.

Also, can make in verbose mode so we can see exactly what compilation command is being passed? It looks to me like you’re doing things correctly.

**Comment by BlackPowerade** [comment]
Found the issue. For some reason     
https://github.com/trzy/Supermodel/blob/master/Makefiles/Makefile.Win32#L96 `$(addprefix -I,$(sort $(PLATFORM_INCLUDE_DIR)))`   
Wasn't adding the paths to the compiler. I have them hard coded (for now) and now it has an issue with linking, but I think I can fix it. Will update later.

**Comment by BlackPowerade** [comment]
Running into issue with unicode and minGW's lack of some kind of support.    
This https://stackoverflow.com/questions/58324230/undefined-reference-to-winmain-c-mingw says you only need to add `-municode` to the compiler options, but in doing so it now produces different errors about converting char* to LPCWSTR:
![image](https://user-images.githubusercontent.com/21957509/189788604-337ff720-93e9-4062-89bc-bf9ab45485e4.png)
I'll keep going at it.

**Comment by ToBul** [comment]
FWIW, I tried this a while ago and also found the GitHub VM file paths way too funky. There, but not there???
I did eventually get it to build a windows exe and upload an 'Artifact' though. I ended up making the sdl.h #includes compiler dependent so the paths would be correct.
I *think* it's only MSVC that has 'SDL.h', while MSYS2, Linux and Mac have 'SDL2/SDL.h'. I might be wrong.
The source would still build locally on Windows in gcc and MSVC. It also built in an Ubuntu VirtualBox with gcc. Wouldn't know about Mac.

Here's a fresh fork,
 https://github.com/ToBul/Supermodel/commit/296181f70f5e3a77bc47cd53349a831e8206af90

**Comment by BlackPowerade** [comment]
More news.
It took me some time before I finally remembered paths were case sensitive on *nix, but when I did Supermodel pretty much built without any problems. It throws some errors about format overflows, but still compiles and links. I'll take that as a win.    
MacOS is a different story, without `-Wno-error=implicit-function-declaration` it fails in m68kcpu and m86kopnz.
![image](https://user-images.githubusercontent.com/21957509/190304625-a300ccb5-95cf-4031-b797-30493d235b5b.png)   
With it, it fails in linking; complaining about not finding SDL, but that doesn't really make sense when it was able to find SDL.h...     
![image](https://user-images.githubusercontent.com/21957509/190305001-9b62c216-11d5-472d-a614-becb7936f653.png)    
Also thanks @ToBul, I've never used C, I would have never figured out the header stuff.

**Comment by ToBul** [comment]
> It throws some errors about format overflows

I had another go at this.
What looks like overflow errors is when you build with netboard but forget to amend SDL net #includes.
Once you do that it builds Win and Ubuntu cleanly and you can leave the Warning settings alone.
Hopefully this will go some way to helping OSX build too. ??

> Also thanks @ToBul, I've never used C

No problem, but I'm no dev or even a coder. Hopefully Bart can review this and make a proper judgement on how to strengthen the #include conditions to suit both real and virtual environments. I know he's busy with real life stuff so this is currently the least of his worries. I wish him all the best and every success.

my 2nd attempt:
https://github.com/ToBul/Supermodel/commit/ca8aad9a82691154f22c7cef8608b416813cb9f7

**Comment by RavenMacDaddy** [comment]
Hey @ToBul, give yourself some credit.

I only know a bit about making manifests and barely see any logic in programming – that's why I wanted to put the right people in contact. :)

**Comment by ToBul** [comment]
I just watched this guy...
https://www.youtube.com/watch?v=V6ACATpl2LQ
I had no idea Apple directory structure was so messy. Looks like getting Supermodel to compile for all systems in both real and virtual environments will be a real pain.

**Comment by trzy** [comment]
It depends how you set the include directory. On Mac it’s SDL2/SDL.h because that’s how the framework directory bundles it. On Windows and Linux you can specify whether to include the SDL2 subfolder or not. 

As for the character set errors, this is very strange. The default options should just work. I haven’t updated my version of gcc and MSYS2 for over a year, though. Is MSYS2 being used here?

**Comment by BlackPowerade** [comment]
Yes, MSYS2 is being used. We can get supermodel to build on windows and linux in actions.    
I've given up on getting MacOS builds for now though. I don't have a Mac to test it on, and I can't really replicate actions environment with docker.

**Comment by ToBul** [comment]
Apologies for necro posting. I had a go at getting Mac to build on git.
Only problem is, I'm not a Mac guy. I have no way of testing the resulting Mac build. 🤣
If someone wants to test this, let me know if it worked. I'll take it down ASAP if it's bad.

https://github.com/ToBul/Supermodel/commits/workflow

It creates 'artifacts' in the actions summaries. 
I don't like the idea of making it post a 'Release'. If we all started posting releases, there would be dodgy versions of Supermodel all over the internet. Bart started official releases on the website to avoid exactly that.

The problems for Mac builds....

```
ld: framework not found SDL2
```
Not sure if brew installs an actual 'Framework', if it does, the symlink stuff might be getting borked. 
Using the more standard lib names in the makefile seems to work.

```
implicit declaration of function 'M68KIRQCallback' is invalid in C99;
```
Mac users have been dealing with this for a couple of years on the Supermodel forums.
Would be cool if it was fixed for them. 😄

The problem seems to be the externally declared functions in,
_Src\CPU\68K\68K.h_,
that are then being defined elsewhere. Modern clang not liking that.

Also, SDL paths as mention in above comments.
I think we're mistaken thinking git runners are like traditional VMs. They're not really trying to replicate an actual environment.
I guess they're just glorified bash shells, grinding through jobs as efficiently as possible. So not surprising the paths and symlink stuff might be a bit funky.

**Comment by ToBul** [comment]
Edit:
It should be said I had changed the SDL includes to get Ubuntu and Win to build.
I didn't try building Mac with the original include conditions.
